### PR TITLE
dojson: more fixes to ids

### PR DESCRIPTION
**Pull request by jacquerie**
Sentry: https://sentry.cern.ch/inspire-sentry/inspire-labs-qa/group/602869/

~~Needs an answer to https://github.com/inspirehep/inspire/issues/211 in order to know which ID schemes we want to implement.~~

- [x] Don't produce incomplete ids
- [x] Recognize malformed CERN ids
- [x] Support GoogleScholar ids
- [x] Support VIAF ids
- [x] Support ResearcherId ids
- [x] Support SCOPUS ids
- [x] ~~Support SLAC ids~~ We already supported them...
- [x] Correct record with SPIRES
- [x] ~~Correct record with TWITTER~~ Under debate in https://github.com/inspirehep/inspire/issues/211#issuecomment-242202346

### CODE REVIEW

File: inspirehep/dojson/hepnames/fields/bd1xx.py [hunk 1]
```diff
@@ -40,6 +40,7 @@
 
 
 INSPIRE_BAI = re.compile('(\w+\.)+\d+')
+LOOKS_LIKE_CERN = re.compile('^\d+$|^CER[MN]|^CNER-|^CVERN', re.I)
```
Comment (kaplun) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: Seriously :scream_cat:

Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: hahaha

Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: It's actually quite accurate though, and very readable ;)

File: tests/unit/dojson/test_dojson_hepnames.py [hunk 2]
```diff
@@ -155,7 +155,7 @@ def test_ids_from_035__a_9_with_cern():
     assert expected == result['ids']
 
 
-def test_ids_from_035__a_9_with_malformed_cern():
+def test_ids_from_035__a_9_with_cern_cern():
```
Comment (david-caro) [inline tests/unit/dojson/test_dojson_hepnames.py]: was this change intended? (now it does not make sense to me, from an out of context perspective, maybe I'm missing something)

File: tests/unit/dojson/test_dojson_hepnames.py [hunk 3]
```diff
@@ -174,6 +174,82 @@ def test_ids_from_035__a_9_with_malformed_cern():
...
+            'type': 'CERN',
+            'value': 'CERN-724319',
+        },
+    ]
+    result = clean_record(hepnames.do(create_record(snippet)))
+
+    assert expected == result['ids']
+
+
+def test_ids_from_035__a_9_with_cner():
```
Comment (david-caro) [inline tests/unit/dojson/test_dojson_hepnames.py]: s/cner/cern

Comment (david-caro) [inline tests/unit/dojson/test_dojson_hepnames.py]: wait, no, it's cner xd

Comment (david-caro) [inline tests/unit/dojson/test_dojson_hepnames.py]: Maybe mix them all into one test with a fixture for the 'misspelled cern name'?

File: inspirehep/dojson/hepnames/fields/bd1xx.py (hunk 1 above)
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: ...and I'm not even done! Look at this beauty:
```XML
<datafield tag="035" ind1=" " ind2=" ">
  <subfield code="9">INSPIRE</subfield>
  <subfield code="a">CER-749663</subfield>
</datafield>
```
(https://inspirehep.net/record/1068076/export/xme)

File: tests/unit/dojson/test_dojson_hepnames.py (hunk 3 above)
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_hepnames.py]: Yeah, when I write code I usually add a test at a time, which is why I end up with so many. I will refactor these into one.

**Comment by jacquerie** [comment]
This will also target https://sentry.cern.ch/inspire-sentry/inspire-labs-qa/group/602871/.

File: tests/unit/dojson/test_dojson_hepnames.py (hunk 3 above)
Comment (jacquerie) [inline tests/unit/dojson/test_dojson_hepnames.py]: Consolidated all malformed CERN tests into one.

**Comment by jacquerie** [comment]
Ready for review.

File: inspirehep/dojson/hepnames/fields/bd1xx.py [hunk 4]
```diff
@@ -143,10 +144,15 @@ def _get_type(value):
             'BAI': 'INSPIRE BAI',
             'CERN': 'CERN',
             'DESY': 'DESY',
+            'GOOGLESCHOLAR': 'GOOGLESCHOLAR',
             'INSPIRE': 'INSPIRE ID',
             'KAKEN': 'KAKEN',
             'ORCID': 'ORCID',
+            'RESEARCHID': 'RESEARCHERID',
+            'RESEARCHERID': 'RESEARCHERID',
```
Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: these should be the same?

Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: I mean the values of `RESEARCHID` and `RESEARCHERID` :)

File: inspirehep/dojson/hepnames/fields/bd1xx.py [hunk 5]
```diff
@@ -159,21 +165,27 @@ def _guess_type_from_value(a_value):
         if INSPIRE_BAI.match(a_value):
             return 'INSPIRE BAI'
 
+    def _try_to_correct_value(_type, a_value):
```
Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: hmm, the `_type` var does not seem to be used, but you use `type_`, a typo?

Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: hmm... probably you wanted to write `a_type` no?

File: inspirehep/dojson/hepnames/fields/bd1xx.py [hunk 6]
```diff
@@ -159,21 +165,27 @@ def _guess_type_from_value(a_value):
...
+        elif type_ == 'KAKEN':
+            return 'KAKEN-' + a_value
+        else:
+            return a_value
+
     a_value = force_single_element(value.get('a'))
 
     type_ = _get_type(value)
     if type_ is None:
         type_ = _guess_type_from_value(a_value)
```
Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: Hmm, this if might be better in the `_try_to_correct_value`.

File: inspirehep/dojson/hepnames/fields/bd1xx.py (hunk 4 above)
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: Nope, this is working around a misspelling in the data: https://github.com/inspirehep/inspire/issues/211#issue-172317903. Actually I should add the ones about `ORCID`.

File: inspirehep/dojson/hepnames/fields/bd1xx.py (hunk 5 above)
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: No, I meant to write `type_`, but Python scoping rules were saving my ass.

File: inspirehep/dojson/hepnames/fields/bd1xx.py (hunk 6 above)
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: You think? The algorithm is:

- Get the type from `035__9`
- If it didn't work try to guess the type from the value in `035__a`
- Given the type and the value, try to massage the value so that it respects the type
- If you found both a type and a value, return them

The general idea is that we trust types more than values.

Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: Hmm, then maybe the `_guess_type_from_value` call should be inside the `_get_type` function.

Comment (david-caro) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: btw. not a big deal at all.

File: inspirehep/dojson/hepnames/fields/bd1xx.py (hunk 4 above)
Comment (jacquerie) [inline inspirehep/dojson/hepnames/fields/bd1xx.py]: Actually I manually corrected the ORCID ones.

**Comment by jacquerie** [comment]
I deleted the Sentry entries abouts `ids` so that we get fresh errors, since the previous ones would have drowned the new ones. This is not something I recommend to do in general (better mark them as solved).
